### PR TITLE
[core] Support dynamic generators for actor tasks

### LINKED ISSUE #28681: [core] Support num_returns="dynamic" generators for actor tasks
### Description

Support [generators](https://docs.ray.io/en/master/ray-core/tasks/generators.html#num-returns-set-by-the-task-executor) for actor tasks.

### Use case

Useful for `compute="actors"` and block splitting in Datasets.

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Support dynamic generators for actor tasks. The main change needed is just to allow num_returns=-1 ("dynamic") for actors.

## Related issue number

Closes #28681.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]
Well, that was easy! Lgtm

File: python/ray/tests/test_generators.py [hunk 1]
```diff
@@ -183,15 +183,36 @@ def generator(num_returns, store_in_plasma, counter):
...
 @pytest.mark.parametrize("store_in_plasma", [False, True])
-def test_dynamic_generator(ray_start_regular, store_in_plasma):
-    @ray.remote(num_returns="dynamic")
-    def dynamic_generator(num_returns, store_in_plasma):
-        for i in range(num_returns):
-            if store_in_plasma:
-                yield np.ones(1_000_000, dtype=np.int8) * i
-            else:
-                yield [i]
+def test_dynamic_generator(ray_start_regular, use_actors, store_in_plasma):
```
Comment (c21) [inline python/ray/tests/test_generators.py:188]: shall we also test failure case, similar to `CustomException` above?

**Review by c21 — APPROVED** [review]
LGTM w/ minor request for unit test. Thanks @stephanie-wang!

File: python/ray/tests/test_generators.py (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/tests/test_generators.py:188]: Sure, will add!

**Comment by stephanie-wang** [comment]
> Well, that was easy! Lgtm

Yeah, most of the work was in #31614 :)

File: python/ray/tests/test_generators.py (hunk 1 above)
Comment (stephanie-wang) [reply python/ray/tests/test_generators.py:188]: Hmm appears that `retry_exception` option isn't supported for actors right now, but I added a test for normal exception handling.

**Review by clarkzinzow — APPROVED** [review]
Nice! I guess removing the dummy object ID nicely aligned the `num_returns` handling for actor tasks and normal tasks, eh? 😁

File: python/ray/tests/test_generators.py (hunk 1 above)
Comment (EricSteinberger) [reply python/ray/tests/test_generators.py:215]: use_actors seems to kick in the branch that doesn't use actors, and vice versa

**Pull request merged**